helix-view/src/document.rs

```diff
@@ -397,30 +398,111 @@ impl Document {
 
     /// The same as [`format`], but only returns formatting changes if auto-formatting
     /// is configured.
-    pub fn auto_format(&self) -> Option<impl Future<Output = LspFormatting> + 'static> {
-        if self.language_config()?.auto_format {
-            self.format()
+    pub fn auto_format(
+        &mut self,
+        view_id: ViewId,
+    ) -> anyhow::Result<Option<impl Future<Output = LspFormatting> + 'static>> {
+        let config = match self.language_config() {
+            Some(config) => config,
+            None => return Ok(None),
+        };
+
+        if config.auto_format {
+            self.format(view_id)
         } else {
-            None
+            Ok(None)
         }
     }
 
     /// If supported, returns the changes that should be applied to this document in order
     /// to format it nicely.
-    pub fn format(&self) -> Option<impl Future<Output = LspFormatting> + 'static> {
-        let language_server = self.language_server()?;
+    pub fn format(
+        &mut self,
+        view_id: ViewId,
+    ) -> anyhow::Result<Option<impl Future<Output = LspFormatting> + 'static>> {
```

Comment:
Oh earlier when I mentioned using Result I meant _replacing_ Option with Result, not wrapping it with Result.

Reply:
Yeah, but then how do I distinguish between, don't return a future because the function failed vs don't return a future because I have a custom formatter? Would I return an error for executing the formatter correctly? What if there is no formatter at all?

Reply:
We can check what type of error is being returned from the result, and propagate it if it's an actual error.